// @subql/validator/dist/readers/github-reader.js
"use strict";
// Copyright 2020-2021 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GithubReader = void 0;
const axios_1 = __importDefault(require("axios"));
const js_yaml_1 = __importDefault(require("js-yaml"));
class GithubReader {
    constructor(key) {
        this.key = key;
        this.api = axios_1.default.create({
            baseURL: `https://raw.githubusercontent.com/${key}`,
        });
    }
    async getPkg() {
        try {
            const branch = await this.getDefaultBranch();
            const { data } = await this.api.get(`${branch}/package.json`);
            return data;
        }
        catch (err) {
            return undefined;
        }
    }
    async getProjectSchema() {
        try {
            const branch = await this.getDefaultBranch();
            const { data } = await this.api.get(`${branch}/project.yaml`);
            return js_yaml_1.default.load(data);
        }
        catch (err) {
            return undefined;
        }
    }
    async getDefaultBranch() {
        if (this.defaultBranch) {
            return this.defaultBranch;
        }
        const { data } = await axios_1.default.get(`https://api.github.com/repos/${this.key}`);
        this.defaultBranch = data.default_branch;
        return this.defaultBranch;
    }
}
exports.GithubReader = GithubReader;
